File: src/io/ziheng/list/leetcode/MergeTwoSortedLists.java
package io.ziheng.list.leetcode;

import io.ziheng.list.leetcode.ListNode;

/**
 * LeetCodde 21. Merge Two Sorted Lists
 * https://leetcode.com/problems/merge-two-sorted-lists/
 */
public class MergeTwoSortedLists {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode dummyHead = new ListNode(-1);
        ListNode currentNode = dummyHead;
        ListNode pA = l1;
        ListNode pB = l2;
        while (pA != null && pB != null) {
            if (pA.val < pB.val) {
                currentNode.next = pA;
                pA = pA.next;
            } else {
                currentNode.next = pB;
                pB = pB.next;
            }
            currentNode = currentNode.next;
        }
        while (pA != null) {
            currentNode.next = pA;
            pA = pA.next;
            currentNode = currentNode.next;
        }
        while (pB != null) {
            currentNode.next = pB;
            pB = pB.next;
            currentNode = currentNode.next;
        }
        return dummyHead.next;
    }
}
/* EOF */

File: src/cc/linkedlist/MergeTwoSortedLists.cc
#include <cstdio>

/**
 * Definition for singly-linked list.
 */
class ListNode {
public:
    int val;
    ListNode *next;
    ListNode(): val(0), next(NULL) {
    }
    ListNode(int x): val(x), next(NULL) {
    }
    ListNode(int x, ListNode *next): val(x), next(next) {
    }
};

/**
 * LeetCode 21. Merge Two Sorted Lists
 * https://leetcode.com/problems/merge-two-sorted-lists/
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode *l1, ListNode *l2) {
        ListNode *dummyHead = new ListNode();
        ListNode *pA = l1;
        ListNode *pB = l2;
        ListNode *currentNode = dummyHead;
        while (pA != NULL && pB != NULL) {
            if (pA -> val < pB -> val) {
                currentNode -> next = pA;
                currentNode = currentNode -> next;
                pA = pA -> next;
            } else {
                currentNode -> next = pB;
                currentNode = currentNode -> next;
                pB = pB -> next;
            }
        }
        while (pA != NULL) {
            currentNode -> next = pA;
            currentNode = currentNode -> next;
            pA = pA -> next;
        }
        while (pB != NULL) {
            currentNode -> next = pB;
            currentNode = currentNode -> next;
            pB = pB -> next;
        }
        ListNode *retNode = dummyHead -> next;
        delete dummyHead;
        return retNode;
    }
};

int main(int argc, char const *argv[]) {
    // ...
    return 0;
}
/* EOF */